selectors/selectors.md

```diff
@@ -302,67 +265,152 @@ If it's a file in some directory, you can also start at a deeper level:
 }}}}}}
 ```
 
-IPLD Schema
------------
+Schema
+------
+
+This code block describes selectors using [IPLD Schemas](../schema-layer/schemas) syntax.
 
 ```ipldsch
-# SelectorComplex is a grouping for most of the selectors, and is really just
-#  a helper for the Selector type, which introduces one more important selector
-#   to the group, but uses a different union representation to do so (for the
-#    sake of overall serial terseness).
-type SelectorComplex union {
-	| SelectAll "a"
-	| SelectFields "f" # n.b. "SelectPath" no longer exists: is a degenerate case of SelectFields.
-	| SelectIndex "i"
-	| SelectRange "r"
-	| SelectRecursive "A"
-	| SelectUnion "u"
+## SelectorEnvelope is the recommended top-level value for serialized messages
+## that don't have established existing context with marks the start of a selector:
+## it's a single-member union used to kick us towards "nominative typing".
+##
+## See https://github.com/ipld/go-ipld-prime/blob/0692e3b8cd7f231fe5d9d16a103bbbacb23dbdb5/doc/schema.md#using-schema-match-checking-as-version-detection
+## for a background on the theory behind this gentle-nominative concept.
+type SelectorEnvelope union {
+	| Selector "selector"
 } representation keyed
 
 type Selector union {
-	| SelectorComplex map
-	| SelectTrue bool
-} representation kinded
-
-# SelectTrue is a terminal selector which will not explore any more fields,
-# and marks the node it's applied to as a match.
-type SelectTrue bool
+	| Matcher "."
+	| ExploreAll "a"
+	| ExploreFields "f" # note "ExplorePath" is a degenerate case of ExploreFields.
+	| ExploreIndex "i"
+	| ExploreRange "r"
+	| ExploreRecursive "R"
+	| ExploreUnion "|"
+	| ExploreConditional "&"
+	| ExploreRecursiveEdge "@" # sentinel value; only valid in some positions.
+} representation keyed
 
-# SelectAll considers all members of the data; think of it as similar to `./*`.
-type SelectAll struct {
-	next Selector (alias ":")
+## ExploreAll is similar to a `*` -- it traverses all elements of an array,
+## or all entries in a map, and applies a next selector to the reached nodes.
+type ExploreAll struct {
+	next Selector (alias ">")
 }
 
-# SelectFields will explore each of the fields listed, and yields a distinct
-# selector for continuing to explore and match on each reached node.
-type SelectFields map {String:Selector}
+## ExploreFields traverses named fields in a map (or equivalently, struct, if
+## traversing on typed/schema nodes) and applies a next selector to the
+## reached nodes.
+##
+## Note that a concept of "ExplorePath" (e.g. "foo/bar/baz") can be represented
+## as a set of three nexted ExploreFields selectors, each specifying one field.
+## (For this reason, we don't have a special "ExplorePath" feature; use this.)
+type ExploreFields struct {
+	fields {String:Selector} (alias "f>")
```

Comment:
Question: The current go-ipld-prime code treats the boolean true as valid in the place of Selector here. Does that mean:
a) ExploreFields should be something like:
```
type ExploreFieldsValue union { boolean | Selector }
type ExploreFields struct {
   fields {String:ExploreFieldsValue} (alias "f>")
}
```
or
b) is the value "true" a valid abbrievation for a selector that is a position only matcher with no label?
(I kind of like the that)

Reply:
That code is old and we should replace it with something matching this schema.

(I kinda liked the "true" as an abbreviation too, tbh, but it made a bunch of other things harder and weirder.  (I have an open TODO to make schemas have a "unit" kind (!) plus representation options so it can declare things such as `type Dummy unit representation true`.  And then make kinded unions know how to support that.  Yeeeee...ahh.))